LIPS/lip-2.md

```diff
@@ -0,0 +1,265 @@
+  `lastCompletedEpochId`. Usually, it should be a frame long: 32 * 12 * 225 = 86400, but maybe
+  multiples more in case that the previous frame didn't reach the quorum.
+
+:warning: It is important to note here, that we collect post/pre pair (not current/last), to avoid
+the influence of new staking during the epoch.
+
+The following contract storage variables are used to keep the information.
+
+    bytes32 internal constant POST_COMPLETED_TOTAL_POOLED_ETHER_POSITION =
+        keccak256("lido.LidoOracle.postCompletedTotalPooledEther");
+    bytes32 internal constant PRE_COMPLETED_TOTAL_POOLED_ETHER_POSITION =
+        keccak256("lido.LidoOracle.preCompletedTotalPooledEther");
+    bytes32 internal constant LAST_COMPLETED_EPOCH_ID_POSITION =
+        keccak256("lido.LidoOracle.lastCompletedEpochId");
+    bytes32 internal constant TIME_ELAPSED_POSITION =
+        keccak256("lido.LidoOracle.timeElapsed");
+
+Public function was added to provide data for calculating the rewards of [stETH][2] holders.
+
+    function getLastCompletedReportDelta()
+        public view
+        returns (
+            uint256 postTotalPooledEther,
+            uint256 preTotalPooledEther,
+            uint256 timeElapsed
+        )
+
+
+## Sanity checks the oracles reports by configurable values.
+
+In order to limit the misbehaving oracles impact, we want to limit oracles report change by 10% APR
+increase in stake and 5% decrease in stake. Both values are configurable by the governance in case of
+extremely unusual circumstances.
+
+:warning: Note that the change is evaluated after the quorum of oracles reports is reached, and not
+on the individual report.
+
+For this, we have added the following accessors and mutators:
+
+    function getAllowedBeaconBalanceAnnualRelativeIncrease() public view returns(uint256)
+
+    function getAllowedBeaconBalanceRelativeDecrease() public view returns(uint256)
+
+    function setAllowedBeaconBalanceAnnualRelativeIncrease(uint256 value)
+        public auth(SET_REPORT_BOUNDARIES)
+
+    function setAllowedBeaconBalanceRelativeDecrease(uint256 value)
+        public auth(SET_REPORT_BOUNDARIES)
+
+And the logic of reporting to the Lido contract got a call to `_reportSanityChecks` that does the
+following. It compares the `preTotalPooledEther` and `postTotalPooledEther` (see above) and
+
+* if there is a profit or same, calculates the [APR][1], compares it with the upper bound. If was above,
+  reverts the transaction with `ALLOWED_BEACON_BALANCE_INCREASE` code.
+* if there is a loss, calculates relative decrease and compares it with the lower bound. If was
+  below, reverts the transaction with `ALLOWED_BEACON_BALANCE_DECREASE` code.
+
+
+## Callback function to be invoked on report pushes.
+
+To provide the external contract with updates on report pushes (every time the quorum is reached
+among oracle daemons data), we provide the following setter and getter functions.
```

Comment:
```suggestion
among oracle daemons data), we provide the following setter and getter functions. It might be needed to implement some updates to the external contracts that should happen at the same tx the rebase happens (e.g. adjusting uniswap v2 pools to reflect the rebase).
```